### PR TITLE
fix(medusa): Amount not handled properly in PaymentProviderService

### PR DESCRIPTION
**Pull request by adrien2p**
**What**

The amount at this stage should not be a constraint to a flag as it is retrieved by either using the cart or the data.
Also, the creation of the intent to the external provider is now properly managed. It means that we should not see more intent than the exact number that we expect.
The customer is now created only once in the external provider.

FIXES CORE-942

### CODE REVIEW

**Comment by adrien2p** [comment]
snapshot-this

**Comment by chemicalkosek** [comment]
After installing this snapshot, I'm getting such error on backend start:
```bash
⠋ Initializing defaults
QueryFailedError: current transaction is aborted, commands ignored until end of transaction block
    at QueryFailedError.TypeORMError [as constructor] (/home/maciek/Dokumenty/websites/wuruki/backend/node_modules/typeorm/error/TypeORMError.js:9:28)
    at new QueryFailedError (/home/maciek/Dokumenty/websites/wuruki/backend/node_modules/typeorm/error/QueryFailedError.js:13:28)
    at PostgresQueryRunner.<anonymous> (/home/maciek/Dokumenty/websites/wuruki/backend/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:281:31)
    at step (/home/maciek/Dokumenty/websites/wuruki/backend/node_modules/tslib/tslib.js:144:27)
    at Object.throw (/home/maciek/Dokumenty/websites/wuruki/backend/node_modules/tslib/tslib.js:125:57)
    at rejected (/home/maciek/Dokumenty/websites/wuruki/backend/node_modules/tslib/tslib.js:116:69)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  query: 'SELECT "Currency"."code" AS "Currency_code", "Currency"."symbol" AS "Currency_symbol", "Currency"."symbol_native" AS "Currency_symbol_native", "Currency"."name" AS "Currency_name" FROM "currency" "Currency" WHERE "Currency"."code" = $1 LIMIT 1',
  parameters: [ 'usd' ],
  driverError: error: current transaction is aborted, commands ignored until end of transaction block
  ```

**Comment by olivermrbl** [comment]
@chemicalkosek Is this locally? Seems completely unrelated to the fix in this PR

**Comment by chemicalkosek** [comment]
@olivermrbl  No, still the same even after removing package-lock. Maybe some other commits from develop branch are in this snapshot and it breaks?

Because it also errors if I install this latest `next` from npm: 1.7.3-next-20230102092215

**Comment by adrien2p** [comment]
I ll have a deeper look but it seams unrelated with that pr as it looks like it occurs from the loaders

**Comment by chemicalkosek** [comment]
Yeah, it needed migrations since there are some location commits in this snapshot:
```bash
14:22 $ medusa migrations run
query: SELECT * FROM current_schema()
query: SHOW server_version;
query: SELECT * FROM "information_schema"."tables" WHERE "table_schema" = 'public' AND "table_name" = 'migrations'
query: SELECT * FROM "migrations" "migrations" ORDER BY "id" DESC
query: START TRANSACTION
query: CREATE TABLE "sales_channel_location" ("id" character varying NOT NULL, "created_at" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "updated_at" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "sales_channel_id" text NOT NULL, "location_id" text NOT NULL, CONSTRAINT "PK_afd2c2c52634bc8280a9c9ee533" PRIMARY KEY ("id"))
query: CREATE INDEX "IDX_6caaa358f12ed0b846f00e2dcd" ON "sales_channel_location" ("sales_channel_id") 
query: CREATE INDEX "IDX_c2203162ca946a71aeb98390b0" ON "sales_channel_location" ("location_id") 
query: CREATE TABLE "product_variant_inventory_item" ("id" character varying NOT NULL, "created_at" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "updated_at" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "inventory_item_id" text NOT NULL, "variant_id" text NOT NULL, "quantity" integer NOT NULL DEFAULT '1', CONSTRAINT "UQ_c9be7c1b11a1a729eb51d1b6bca" UNIQUE ("variant_id", "inventory_item_id"), CONSTRAINT "PK_9a1188b8d36f4d198303b4f7efa" PRIMARY KEY ("id"))
query: CREATE INDEX "IDX_c74e8c2835094a37dead376a3b" ON "product_variant_inventory_item" ("inventory_item_id") 
query: CREATE INDEX "IDX_bf5386e7f2acc460adbf96d6f3" ON "product_variant_inventory_item" ("variant_id") 
query: ALTER TABLE "return" ADD "location_id" character varying
query: ALTER TABLE "store" ADD "default_location_id" character varying
query: INSERT INTO "migrations"("timestamp", "name") VALUES ($1, $2) -- PARAMETERS: [1666251508718,"multiLocation1666251508718"]
query: COMMIT
info:    Migrations completed.
```

**Comment by chemicalkosek** [comment]
Still seeing the same issue. Supposed to be three intents. This is two carts:
image

**Comment by adrien2p** [comment]
@chemicalkosek are you sure you are running the snap? cause you can see the result I get here https://github.com/medusajs/medusa/issues/2736#issuecomment-1368880430

My flow is
- Create a cart with items
- Update address and email
- call payment-sessions
- Set stripe as payment session
- add a new item to the cart

**Comment by adrien2p** [comment]
this is the snapshot to install -> @medusajs/medusa@1.7.3-snapshot-20230102120149

**Comment by adrien2p** [comment]
Screenshot 2023-01-02 at 15 25 18
Screenshot 2023-01-02 at 15 26 14

I just re-tested it and I get the right call, one customer is created, then one intent, and two updates to the previously created intent to update the amount. Also, we can see that I have one payment per customer

**Comment by chemicalkosek** [comment]
The original issue (https://github.com/medusajs/medusa/issues/2736) states, that you need at least two of Stripe providers enabled in region settings.
Indeed it works correctly if there's just Stripe. It breaks when there's two or more.

**Comment by chemicalkosek** [comment]
Yeah, tested multiple payments on the new snapshot and now it all works superb :+1: 

Apart from the expected higher response times on cart manipulation as discussed on discord :)

**Comment by adrien2p** [comment]
Screenshot 2023-01-03 at 10 21 28
Screenshot 2023-01-03 at 10 21 48

File: packages/medusa/src/services/payment-provider.ts [hunk 1]
```diff
@@ -280,21 +281,30 @@ export default class PaymentProviderService extends TransactionBaseService {
       const provider = this.retrieveProvider(paymentSession.provider_id)
 
       const context = this.buildPaymentContext(sessionInput)
+      context.customer = context.customer?.id
```
Comment (riqwan) [inline packages/medusa/src/services/payment-provider.ts:284]: suggestion: this can potentially be moved to a util function.

**Review by riqwan — COMMENTED** [review]
looks good, added some non-blocking comments/questions :)

File: packages/medusa/src/interfaces/payment-service.ts [hunk 2]
```diff
@@ -130,7 +130,7 @@ export abstract class AbstractPaymentService
   public abstract updatePayment(
     paymentSessionData: PaymentSessionData,
     context: Cart & PaymentContext
-  ): Promise<PaymentSessionResponse>
+  ): Promise<PaymentSessionResponse | PaymentSessionResponse["session_data"]>
```
[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:133]: Yes it is just for the backward compact 👍

File: packages/medusa/src/services/cart.ts [hunk 3]
```diff
@@ -1746,34 +1746,35 @@ class CartService extends TransactionBaseService {
...
-                const paymentSessionInput = {
-                  ...partialSessionInput,
-                  provider_id: paymentSession.provider_id,
-                }
+          for (const paymentSession of cart.payment_sessions) {
+            if (
+              total <= 0 ||
+              !region.payment_providers.find(
+                ({ id }) => id === paymentSession.provider_id
+              )
```
[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/services/cart.ts:1756]: It has been move in the other pr about that subject 🚀

File: packages/medusa/src/services/cart.ts [hunk 4]
```diff
@@ -1785,29 +1786,26 @@ class CartService extends TransactionBaseService {
...
+                continue
+              }
+
+              const paymentSessionInput = {
+                ...partialSessionInput,
+                provider_id: paymentProvider.id,
+              }
+
+              await paymentProviderServiceTx.createSession(paymentSessionInput)
+            }
```
[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/services/cart.ts:1804]: No, we are in the set payment sessions here. So it create the sessions but does not mark any as selected. setPaymentSession kicks in for that. But the management of the sessions is part of the other pr

File: packages/medusa/src/services/cart.ts [hunk 5]
```diff
@@ -1746,34 +1746,35 @@ class CartService extends TransactionBaseService {
...
+              continue
+            }
 
-                return this.paymentProviderService_
-                  .withTransaction(transactionManager)
-                  .updateSession(paymentSession, paymentSessionInput)
-              }
-            })
-          )
+            seen.push(paymentSession.provider_id)
```
[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/services/cart.ts:1762]: will be part of the session flow pr 👍

**Pull request closed without merging**